File: skript-parser-java/src/main/java/io/github/syst3ms/skriptparser/classes/SkriptParser.java
package io.github.syst3ms.skriptparser.classes;

import io.github.syst3ms.skriptparser.pattern.ChoiceElement;
import io.github.syst3ms.skriptparser.pattern.ChoiceGroup;
import io.github.syst3ms.skriptparser.pattern.CompoundElement;
import io.github.syst3ms.skriptparser.pattern.ExpressionElement;
import io.github.syst3ms.skriptparser.pattern.OptionalGroup;
import io.github.syst3ms.skriptparser.pattern.PatternElement;
import io.github.syst3ms.skriptparser.pattern.RegexGroup;
import io.github.syst3ms.skriptparser.pattern.TextElement;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;

/**
 * A parser instance used for matching a pattern to a syntax
 */
public class SkriptParser {
    private String pattern;
    private PatternElement element;
    private int patternIndex = 0;
    private List<Expression<?>> parsedExpressions = new ArrayList<>();
    private List<MatchResult> regexMatches = new ArrayList<>();
    private int parseMark = 0;

    public SkriptParser(PatternElement e) {
        this.pattern = e.toString();
        element = e;
    }

    public String getPattern() {
        return pattern;
    }

    public PatternElement getElement() {
        return element;
    }

    public int getPatternIndex() {
        return patternIndex;
    }

    public void advanceInPattern() {
        patternIndex++;
    }

    public List<Expression<?>> getParsedExpressions() {
        return parsedExpressions;
    }

    public void addExpression(Expression<?> expression) {
        parsedExpressions.add(expression);
    }

    public List<MatchResult> getRegexMatches() {
        return regexMatches;
    }

    public void addRegexMatch(MatchResult match) {
        regexMatches.add(match);
    }

    public int getParseMark() {
        return parseMark;
    }

    public void addMark(int mark) {
        parseMark ^= mark;
    }

    public ParseResult toParseResult() {
        return new ParseResult(element, regexMatches, parseMark);
    }

    public List<PatternElement> flatten(PatternElement element) {
        if (element instanceof CompoundElement) {
            return ((CompoundElement) element).getElements();
        } else {
            return Collections.singletonList(element);
        }
    }

    public List<PatternElement> getPossibleInputs(List<PatternElement> elements) {
        List<PatternElement> possibilities = new ArrayList<>();
        for (PatternElement element : elements) {
            if (element instanceof TextElement || element instanceof RegexGroup) {
                if (element instanceof TextElement && ((TextElement) element).getText().matches("\\s+"))
                    continue;
                possibilities.add(element);
                return possibilities;
            } else if (element instanceof ChoiceGroup) {
                for (ChoiceElement choice : ((ChoiceGroup) element).getChoices()) {
                    possibilities.addAll(getPossibleInputs(flatten(choice.getElement())));
                }
                return possibilities;
            } else if (element instanceof ExpressionElement) { // Can't do much about this
                possibilities.add(new RegexGroup(Pattern.compile(".+")));
                return possibilities;
            } else if (element instanceof OptionalGroup) {
                possibilities.addAll(getPossibleInputs(flatten(((OptionalGroup) element).getElement())));
            }
        }
        possibilities.add(new TextElement(""));
        return possibilities;
    }

    public Expression<?> parseExpression(String s) { // empty implementation
        if (s.equals("2"))
            return new Literal<>(Number.class, 2);
        return null;
    }
}

File: skript-parser-kotlin/src/main/kotlin/io/github/syst3ms/skriptparser/classes/SkriptParser.kt
package io.github.syst3ms.skriptparser.classes

import io.github.syst3ms.skriptparser.pattern.*

/**
 * A parser instance used for matching a pattern to a syntax, stores a parse mark
 */
class SkriptParser(val element: PatternElement) {
    val pattern: String = element.toString()
    var patternIndex = 0
        private set
    val parsedExpressions = arrayListOf<Expression<*>>()
    val regexMatches = arrayListOf<MatchResult>()
    var parseMark = 0
        private set

    fun advanceInPattern() {
        patternIndex++
    }

    fun addExpression(expression: Expression<*>) {
        parsedExpressions += expression
    }

    fun addRegexMatch(match: MatchResult) {
        regexMatches.add(match)
    }

    fun addMark(mark: Int) {
        parseMark = parseMark xor mark
    }

    fun flatten(element: PatternElement): List<PatternElement> {
        val flattened = arrayListOf<PatternElement>()
        if (element is CompoundElement) {
            for (e in element.elements) {
                flattened.addAll(flatten(e))
            }
            return flattened
        } else {
            flattened.add(element)
            return flattened
        }
    }

    fun getPossibleInputs(elements: List<PatternElement>): List<PatternElement> {
        val possibilities = arrayListOf<PatternElement>()
        elementLoop@ for (element in elements) {
            when (element) {
                is TextElement, is RegexGroup -> {
                    if (element is TextElement && element.text.isBlank())
                        continue@elementLoop
                    possibilities.add(element)
                    return possibilities
                }
                is ChoiceGroup -> {
                    for ((choice) in element.choices) {
                        possibilities.addAll(getPossibleInputs(flatten(choice)))
                    }
                    return possibilities
                }
                is ExpressionElement -> { // Can't do much about this
                    possibilities.add(RegexGroup(".+".toRegex().toPattern()))
                    return possibilities
                }
                is OptionalGroup -> possibilities.addAll(getPossibleInputs(flatten(element.element)))
            }
        }
        possibilities.add(TextElement(""))
        return possibilities
    }

    fun parseExpression(s: String): Expression<*>? { // empty implementation
        return if (s == "2") object : Expression<Number> {
            override fun hashCode(): Int {
                return 2
            }
        } else null
    }
}
